ulib/FStar.Algebra.CommMonoid.Fold.fst

```diff
@@ -0,0 +1,129 @@
+(*
+   Copyright 2008-2018 Microsoft Research
+   
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+   
+       http://www.apache.org/licenses/LICENSE-2.0
+       
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+
+   Author: A. Rozanov
+*)
+
+module FStar.Algebra.CommMonoid.Fold
+
+open FStar.Seq.Base
+open FStar.Seq.Properties
+open FStar.Seq.Permutation
+
+(* Here we define the notion for big sums and big products for 
+   arbitrary commutative monoids. We construct the folds from 
+   an integer range and a function, then calculate the fold --
+   a sum or a product, depending on the monoid operation. *)
+ 
+(* We refine multiplication a bit to make proofs smoothier *)
+let ( *) (x y: int) : (t:int{(x>0 /\ y>0) ==> t>0}) = op_Multiply x y
```

Comment:
I'm not sure this re-definition of multiplication is worth it. Is it really making proofs simpler? I would just open `FStar.Mul` and use the `*` that it defines.